cdap-app-fabric/src/main/java/io/cdap/cdap/internal/sysapp/SystemAppEnableExecutor.java

```diff
@@ -0,0 +1,141 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.sysapp;
+
+import com.google.common.base.Preconditions;
+import com.google.gson.Gson;
+import com.google.inject.Inject;
+import io.cdap.cdap.api.artifact.ArtifactSummary;
+import io.cdap.cdap.api.dataset.DatasetManagementException;
+import io.cdap.cdap.api.retry.RetryableException;
+import io.cdap.cdap.app.program.ProgramDescriptor;
+import io.cdap.cdap.common.ApplicationNotFoundException;
+import io.cdap.cdap.common.ConflictException;
+import io.cdap.cdap.common.InvalidArtifactException;
+import io.cdap.cdap.common.NotFoundException;
+import io.cdap.cdap.common.service.Retries;
+import io.cdap.cdap.common.service.RetryStrategies;
+import io.cdap.cdap.internal.app.deploy.pipeline.ApplicationWithPrograms;
+import io.cdap.cdap.internal.app.services.ApplicationLifecycleService;
+import io.cdap.cdap.internal.app.services.ProgramLifecycleService;
+import io.cdap.cdap.internal.sysapp.SystemAppStep.Arguments;
+import io.cdap.cdap.proto.ProgramStatus;
+import io.cdap.cdap.proto.id.ApplicationId;
+import io.cdap.cdap.proto.id.KerberosPrincipalId;
+import io.cdap.cdap.proto.id.ProgramId;
+import io.cdap.cdap.security.spi.authorization.UnauthorizedException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Deploys an application with given arguments. Also runs all programs corresponding to the
+ * application.
+ */
+public class SystemAppEnableExecutor {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SystemAppEnableExecutor.class);
+  private static final Gson GSON = new Gson();
+  private final ApplicationLifecycleService appLifecycleService;
+  private final ProgramLifecycleService programLifecycleService;
+
+  @Inject
+  SystemAppEnableExecutor(ApplicationLifecycleService appLifecycleService,
+      ProgramLifecycleService programLifecycleService) {
+    this.appLifecycleService = appLifecycleService;
+    this.programLifecycleService = programLifecycleService;
+  }
+
+  public void deployAppAndStartPrograms(Arguments arguments) {
+    try {
+      // TODO(CDAP-16243): Make deploy app and start programs idempotent to not have transactional issues.
+      ApplicationWithPrograms appDetail = retryableDeploySystemApp(arguments);
+      for (ProgramDescriptor program : appDetail.getPrograms()) {
+        startProgram(program.getProgramId());
```

Comment:
doesn't look like there are any retries if this fails

Reply:
If the program is already running, this will be a no-op. However, this actually isn't the desired behavior if the app changed.

For example, if the artifact version is different, after we deploy the app, we need to restart the programs in order for them to pick up the new code. We need to restart the programs if the artifact or config for the app changed. 

Or is this what you were referring to with that idempotent TODO comment?

Reply:
> doesn't look like there are any retries if this fails

Not retrying here as ProgramStarter does not throw retryable exception. I am only retrying for DeployingSystemApp because it throws retryable exception (and we need to wait till metadata sergice is up and running). Following what we did for ProgramStarter Bootstrap step.

Reply:
> If the program is already running, this will be a no-op. However, this actually isn't the desired behavior if the app changed.
> 
> For example, if the artifact version is different, after we deploy the app, we need to restart the programs in order for them to pick up the new code. We need to restart the programs if the artifact or config for the app changed.
> 
> Or is this what you were referring to with that idempotent TODO comment?

Yeah. I am planning to work on understanding how to handle those cases from you after this change is in and will start working on it after operator changes are also in.

Reply:
The caller decides what it wants to retry, not the ProgramLifecycleService. You can define whatever retry logic you want.

You should not compare this service to the BootstrapService. They have completely different contracts. The contract here is that CDAP will make system apps match the state declared in the configs. If program start fails for some transient reason, it should be retried in order to fulfill that contract. If it failed in a way that we know it will always fail, then it doesn't need to be retried.